roofit/roofitcore/src/RooMinimizer.cxx

```diff
@@ -816,8 +799,61 @@ void RooMinimizer::profileStop()
 }
 
 
+ROOT::Math::IMultiGenFunction* RooMinimizer::getFitterMultiGenFcn() const
+{
+   return fitter()->GetFCN();
+}
+
+
+ROOT::Math::IMultiGenFunction* RooMinimizer::getMultiGenFcn() const
+{
+   if (getFitterMultiGenFcn()) {
+      return getFitterMultiGenFcn();
+   } else {
+      switch (_fcnMode) {
+      case FcnMode::classic: {
+         return static_cast<ROOT::Math::IMultiGenFunction *>(static_cast<RooMinimizerFcn *>(_fcn));
+      }
+      case FcnMode::gradient: {
+         auto thing = dynamic_cast<RooGradMinimizerFcn *>(_fcn);
+         if (thing != nullptr) {
+            return static_cast<ROOT::Math::IMultiGenFunction *>(thing);
+         } else {
+            throw std::logic_error("In RooMinimizer::fitterFcn: Minimizer fcnMode was set to gradient, but fit function type does not match RooGradMinimizerFcn!");
+         }
+      }
+      }
+   }
```

Comment:
```suggestion
      }
      return nullptr;
   }
```
To avoid a warning about not returning anything at the end of a non-void function.

Reply:
Hm, I don't get that warning, is it somehow compiler/platform dependent? Weird. I'm a bit hestitant about returning a null pointer. I think the control flow should not reach that point at all; every block either returns something or throws. Making it return a nullptr would effectively signal some exotic error state that should be solved on the developer side, not the user side. I don't know exactly what the ROOT policy is on exceptions, but I would personally then opt for throwing at this point.

Reply:
I see your point, but I also can't change the fact that I get the following warning with gcc11 :)
```
/home/rembserj/root-builds/master/src/roofit/roofitcore/src/RooMinimizer.cxx: In member function ‘ROOT::Math::IMultiGenFunction* RooMinimizer::getMultiGenFcn() const’:
/home/rembserj/root-builds/master/src/roofit/roofitcore/src/RooMinimizer.cxx:827:1: warning: control reaches end of non-void function [-Wreturn-type]
  827 | }
      | ^
/home/rembserj/root-builds/master/src/roofit/roofitcore/src/RooMinimizer.cxx: In member function ‘const RooAbsMinimizerFcn* RooMinimizer::fitterFcn() const’:
/home/rembserj/root-builds/master/src/roofit/roofitcore/src/RooMinimizer.cxx:849:1: warning: control reaches end of non-void function [-Wreturn-type]
  849 | }
      | ^
```
Maybe we can circumvent the problem by simplifying the relevant functions `getMultiGenFcn` and `fitterFcn`? For example, I `getMultiGenFcn` you are always having a static cast to ROOT::Math::IMultiGenFunction for both the `classic` and `gradient` case. The switch statement was just to catch the logic error that is already checked for in the new RooMinimizer constructor. So would it be possible to do this?
```C++
ROOT::Math::IMultiGenFunction* RooMinimizer::getMultiGenFcn() const
{                           
   if (getFitterMultiGenFcn()) {
      return getFitterMultiGenFcn();
   }
   return static_cast<ROOT::Math::IMultiGenFunction *>(static_cast<RooMinimizerFcn *>(_fcn));
}
```
I think also the logic of `RooMinimizer::fitterFcn()` can be simplified, but I'll make a separate comment for that.

Reply:
Is `static_cast<ROOT::Math::IMultiGenFunction *>(static_cast<RooMinimizerFcn *>(_fcn))` always going to work? For instance, is it guaranteed that every `RooAbsMinimizerFcn` subclass will have the same data member ordering (at least for the `IMultiGenFunction` part) as `RooMinimizerFcn`?

Reply:
Oh yes I see, it's also dangerous in that regard....